src/basic/path-util.c

```diff
@@ -599,14 +600,24 @@ static int check_x_access(const char *path, int *ret_fd) {
                 r = access_fd(fd, X_OK);
                 if (r < 0)
                         return r;
-
-                *ret_fd = TAKE_FD(fd);
         } else {
                 /* Let's optimize things a bit by not opening the file if we don't need the fd. */
                 if (access(path, X_OK) < 0)
                         return -errno;
         }
 
+        with_dash = strjoina(path, "/");
+
+        /* If this passes, it must be a directory. */
+        if (access(with_dash, X_OK) >= 0)
+                return -EISDIR;
+
+        if (errno != ENOTDIR)
+                return -errno;
+
+        if (ret_fd)
+                *ret_fd = TAKE_FD(fd);
+
```

Comment:
hmm, so this now does two access() calls for the case where the fd is not requested, and opens things up for races (theoretically, …). now, the access() code path was in place for optimization, but if an optimization is less sexy than the original (and i think it is less sexy, since more than one syscall, and non-atomic), i think we shouldn't do the optimization.

maybe we should just always use open(O_PATH) now, and then fstat() it to check if regular file, and then access_fd() it for checking if its executable.

```c
fd = open(path, O_CLOEXEC|O_PATH);
if (fd < 0)
        return -errno;

r = fd_verify_regular(fd);
if (r < 0)
        return r;

r = access_fd(fd, X_OK);
if (r < 0)
       return r;

if (ret_fd)
        *ret_fd = TAKE_FD(fd);
```

This would then mean 3 syscalls for cases where the caller wants the fd, and 4 where they don't. which is ok, i guess.